### PR TITLE
refactor: Create memoized tree traversal helpers

### PR DESCRIPTION
**Pull request by TrevorBergeron**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: bigframes/core/nodes.py [hunk 1]
```diff
@@ -1486,3 +1486,56 @@ def remap_refs(
...
+    validate=False,
+) -> BigFrameNode:
+    """
+    Perform a bottom-up transformation of the BigFrameNode tree.
+
+    If memoize=True, recursive calls are memoized within the scope of the traversal only.
+    """
+
+    def bottom_up_internal(root: BigFrameNode) -> BigFrameNode:
+        return transform(root.transform_children(bottom_up_internal))
```
Comment (GarrettWu) [inline bigframes/core/nodes.py:1532]: Don't see a difference with top_down. Should bottom_up apply to parents?

Comment (TrevorBergeron) [reply bigframes/core/nodes.py:1532]: top down applies `transform_children()` after `transform(root)` that is the only difference.

Comment (GarrettWu) [reply bigframes/core/nodes.py:1532]:
```
def bottom_up_internal(root: BigFrameNode) -> BigFrameNode:
        return transform(root.transform_children(bottom_up_internal))
```
Then this is a mistake?

Comment (TrevorBergeron) [reply bigframes/core/nodes.py:1532]: nope, not a mistake. bottom_up transforms the children first. and then transforms the root last.

Comment (GarrettWu) [reply bigframes/core/nodes.py:1532]: Then bottom up doesn't apply to their parents? How is that "bottom up"?

Comment (TrevorBergeron) [reply bigframes/core/nodes.py:1532]: to which parents? its a classic post-order depth-first traversal, all nodes under the root are transformed, and then the root is the final node to be transformed

Comment (GarrettWu) [reply bigframes/core/nodes.py:1532]: OIC. I was thinking it to start from leaf, but it finds the leafs from the root. Cool.

**Review by GarrettWu — APPROVED** [review]

**Pull request merged**